lightly/loss/msn_loss.py

```diff
@@ -0,0 +1,203 @@
+import math
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.distributed as dist
+
+
+def prototype_probabilities(
+    queries: torch.Tensor,
+    prototypes: torch.Tensor,
+    temperature: float,
+) -> torch.Tensor:
+    """Returns probability for each query to belong to each prototype.
+
+    Args:
+        queries:
+            Tensor with shape (batch_size, dim)
+        prototypes:
+            Tensor with shape (num_prototypes, dim)
+        temperature:
+            Inverse scaling factor for the similarity.
+
+    Returns:
+        Probability tensor with shape (batch_size, num_prototypes) which sums to 1 along
+        the num_prototypes dimension.
+
+    """
+    return F.softmax(torch.matmul(queries, prototypes.T) / temperature, dim=1)
+
+def sharpen(probabilities: torch.Tensor, temperature: float) -> torch.Tensor:
+    """Sharpens the probabilities with the given temperature.
+
+    Args:
+        probabilities:
+            Tensor with shape (batch_size, dim)
+        temperature:
+            Temperature in (0, 1]. Lower temperature results in stronger sharpening (
+            output probabilities are less uniform).
+    Returns:
+        Probabilities tensor with shape (batch_size, dim).
+
+    """
+    probabilities = probabilities ** (1.0 / temperature)
+    probabilities /= torch.sum(probabilities, dim=1, keepdim=True)
+    return probabilities
+
+@torch.no_grad()
+def sinkhorn(
```

Comment:
Can we not re-use the sinkhorn vom swav? E.g. move it to utils

Reply:
It is slightly different :(